source/tests/common/dpmodel/test_output_def.py

```diff
@@ -103,6 +106,101 @@ def test_model_output_def(self):
+        self.assertEqual(md["foo"].category & OutputVariableOperation.REDU, 0)
+        self.assertEqual(md["foo"].category & OutputVariableOperation.DERV_R, 0)
+        self.assertEqual(md["foo"].category & OutputVariableOperation.DERV_C, 0)
+        self.assertEqual(
+            md["energy_redu"].category & OutputVariableOperation.REDU,
+            OutputVariableOperation.REDU,
+        )
+        self.assertEqual(md["energy_redu"].category & OutputVariableOperation.DERV_R, 0)
+        self.assertEqual(md["energy_redu"].category & OutputVariableOperation.DERV_C, 0)
+        self.assertEqual(md["energy_derv_r"].category & OutputVariableOperation.REDU, 0)
+        self.assertEqual(
+            md["energy_derv_r"].category & OutputVariableOperation.DERV_R,
+            OutputVariableOperation.DERV_R,
+        )
+        self.assertEqual(
+            md["energy_derv_r"].category & OutputVariableOperation.DERV_C, 0
+        )
+        self.assertEqual(md["energy_derv_c"].category & OutputVariableOperation.REDU, 0)
+        self.assertEqual(
+            md["energy_derv_c"].category & OutputVariableOperation.DERV_R, 0
+        )
+        self.assertEqual(
+            md["energy_derv_c"].category & OutputVariableOperation.DERV_C,
+            OutputVariableOperation.DERV_C,
+        )
+        self.assertEqual(
+            md["energy_derv_c_redu"].category & OutputVariableOperation.REDU,
+            OutputVariableOperation.REDU,
+        )
+        self.assertEqual(
+            md["energy_derv_c_redu"].category & OutputVariableOperation.DERV_R, 0
+        )
+        self.assertEqual(
+            md["energy_derv_c_redu"].category & OutputVariableOperation.DERV_C,
+            OutputVariableOperation.DERV_C,
+        )
+
+        # apply_operation
+        self.assertEqual(
+            apply_operation(md["energy"], OutputVariableOperation.REDU),
+            md["energy_redu"].category,
+        )
+        self.assertEqual(
+            apply_operation(md["energy"], OutputVariableOperation.DERV_R),
+            md["energy_derv_r"].category,
+        )
+        self.assertEqual(
+            apply_operation(md["energy"], OutputVariableOperation.DERV_C),
+            md["energy_derv_c"].category,
+        )
+        self.assertEqual(
+            apply_operation(md["energy_derv_c"], OutputVariableOperation.REDU),
+            md["energy_derv_c_redu"].category,
+        )
+        # raise ValueError
+        with self.assertRaises(ValueError):
+            apply_operation(md["energy_redu"], OutputVariableOperation.REDU)
+        with self.assertRaises(ValueError):
+            apply_operation(md["energy_derv_c"], OutputVariableOperation.DERV_C)
+        with self.assertRaises(ValueError):
+            apply_operation(md["energy_derv_c_redu"], OutputVariableOperation.REDU)
+        # hession
+        hession_cat = apply_operation(
+            md["energy_derv_r"], OutputVariableOperation.DERV_R
```

Comment:
what is the convention of getting hessian?
1. apply_operation(apply_operation(energy, derv_r), derv_r)
2. apply_operation(apply_operation(energy, derv_r), sec_derv_r)

Reply:
(1) is implicit, and (2) is explicit. But even for the explicit case, we still have to check in `do_derivative`.

Reply:
I don't oppose the explicit case...

Reply:
Or, does `do_derivative` only work for the original atomic energy/tensor?

Reply:
The force derivative works for any tensor. while the cell derivative works only on the force does not make sense. 
It seems that we should introduce differentiable_r and differentiable_c instead of differentiable. 
I would prefer the implicit case, the derv_r can be use for both the first and second order derivatives.